### PR TITLE
[SYCLomatic] Do the scaling for alpha/a_scale/b_scale before matmul operation

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 1]
```diff
@@ -222,60 +222,117 @@ class matmul_desc_t {
...
+                                             library_data_t::real_float): {
+    e = scale_new_a_impl<float, float>(
+        q_ptr, rows, cols, (float *)a, (const float *)alpha, vector_alpha,
+        device_alpha, (const float *)a_scale, (const float *)b_scale, deps);
     break;
   }
   default:
-    throw std::runtime_error("dpct::blas_gemm::experimental::detail::scale_d_"
-                             "with_vector_alpha() does not support the data "
+    printf("a_type:%d, scale_type:%d\n", (int)a_type, (int)scale_type);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 2]
```diff
@@ -222,60 +222,145 @@ class matmul_desc_t {
...
+  if (device_alpha) {
+    return q_ptr->submit([&](sycl::handler &cgh) {
+      cgh.depends_on(deps);
+#ifdef DPCT_USM_LEVEL_NONE
+      access_wrapper<T *> a_acc(a, cgh);
+      const Tscale *alpha_ptr = nullptr;
+      if (device_alpha)
+        alpha_ptr = alpha;
+      access_wrapper<const Tscale *> alpha_acc(alpha_ptr, cgh);
+      auto a_scale_buf = a_scale ? dpct::get_buffer(a_scale) : dpct::buffer_t(sycl::range<1>(0));
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: use this way to pass the alpha to device side. Then we can merge the 2 submit calls.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 3]
```diff
@@ -232,60 +232,107 @@ class matmul_desc_t {
...
+  Tscale *a_scale = (Tscale *)a_scale_ptr;
+  Tscale *b_scale = (Tscale *)b_scale_ptr;
+
+  if (!device_alpha) {
+    Tscale *alpha_host = alpha;
+    alpha = (Tscale *)::dpct::cs::malloc(sizeof(Tscale), *q_ptr);
+    deps.push_back(
+        ::dpct::cs::memcpy(*q_ptr, alpha, alpha_host, sizeof(Tscale)));
+  }
+  if (!a_scale_ptr)
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: memory_detail.hpp:210

**Review by ziranzha — APPROVED** [review]

**Pull request merged**